Fix condition checks regarding oozie sharelib dir/lib update

kiiranh opened the pull request:
I observed on a test VM cluster that a new oozie sharelib directory (hdfs:///user/oozie/share/lib/lib_<timestamp>) was created on every chef-client run. This PR fixes a few conditional errors in the code which manages the sharelib directory.

Fixes Issue #178

amithkanand commented on cookbooks/bcpc-hadoop/recipes/oozie.rb:
```diff
@@ -106,23 +106,14 @@
   #not_if { require 'time'
   # hdfs_mtime=`hdfs dfs -stat #{share_dir_url}`.strip
   # Time.parse("#{hdfs_mtime} UTC") >
   # File.mtime("#{OOZIE_CLIENT_PATH}/oozie-sharelib.tar.gz") }
```
Can we remove these commented lines?

kiiranh replied:
Per @cbaenziger, these need to be uncommented again so sharelib is updated on hdfs. The current change will only make sure sharelib dir is created once on hdfs, it won't update if the on-disk sharelib has changed

amithkanand commented on cookbooks/bcpc-hadoop/recipes/oozie.rb:
```diff
@@ -106,23 +106,14 @@
   #code "#{OOZIE_LIB_PATH}/bin/oozie-setup.sh sharelib update -fs #{HDFS_URL}"
```
Can we remove these commented line?

kiiranh commented:
@amithkanand Comments incorporated. Now, sharelib is updated on hdfs only if the sharelib tarball has been changed/updated.

cbaenziger commented on cookbooks/bcpc-hadoop/recipes/oozie.rb:
```diff
@@ -65,61 +56,73 @@
+
+update_oozie_sharelib = lambda do |host|
+  begin
+    oozie_status=`sudo -u oozie oozie admin -oozie http://#{host}:11000/oozie -status 2>&1`
```
Please use [mixlib-shellout](https://github.com/chef/mixlib-shellout) instead of ```

cbaenziger commented on cookbooks/bcpc-hadoop/recipes/oozie.rb:
```diff
@@ -65,61 +56,73 @@
+update_oozie_sharelib = lambda do |host|
+  begin
+    oozie_status=`sudo -u oozie oozie admin -oozie http://#{host}:11000/oozie -status 2>&1`
+    if /NORMAL/ =~ oozie_status && $?.to_i
```
Here `$?.to_i` will always evaluate to true; regardless of if `$?` is 0 or 1. I suspect the intention is `$? != 0`.

http-418 replied:
$? has the methods "success?" and "exitstatus" for this purpose

(as with perl, taking $? directly as an integer can have surprising results, because the exit code is stored in the high bits of a 16 bit integer)

cbaenziger commented on cookbooks/bcpc-hadoop/recipes/oozie.rb:
```diff
@@ -65,61 +56,73 @@
+    oozie_status=`sudo -u oozie oozie admin -oozie http://#{host}:11000/oozie -status 2>&1`
+    if /NORMAL/ =~ oozie_status && $?.to_i
+      update_status=`sudo -u oozie oozie admin -oozie http://#{host}:11000/oozie -sharelibupdate 2>&1`
+      Chef::Log.warn("Sharelibupdate: Updated sharelib on #{host}") if $?.to_i
```
Again `$? != 0`?

cbaenziger commented on cookbooks/bcpc-hadoop/recipes/oozie.rb:
```diff
@@ -65,61 +56,73 @@
+ruby_block "notify_sharelib_update" do
+  block do
+    node[:bcpc][:hadoop][:oozie_hosts].each do |oozie_host|
+      update_oozie_sharelib.call(oozie_host[:hostname])
```
As you pointed out; should be `float_host()`

cbaenziger commented on cookbooks/bcpc-hadoop/recipes/oozie.rb:
```diff
@@ -155,15 +158,14 @@
         db.write code
       end
-      #system "sudo -u oozie #{OOZIE_CLIENT_PATH}/bin/ooziedb.sh create -sqlfile #{OOZIE_CLIENT_PATH}/oozie.sql -run Validate DB Connection"
       system "sudo -u oozie /usr/hdp/current/oozie-server/bin/ooziedb.sh create -sqlfile /usr/hdp/current/oozie-server/oozie.sql -run Validate DB Connection"
```
Could you continue to use `#{OOZIE_CLIENT_PATH}`?

kiiranh replied:
Can't use `#{OOZIE_CLIENT_PATH}` since the file `ooziedb.sh` moved from that path to `/usr/hdp/current/oozie-server/bin/ooziedb.sh`:
```bash
bcpc-vm2:~$ find /usr/hdp -type f -name ooziedb*
/usr/hdp/2.2.0.0-2041/oozie/bin/ooziedb.sh
bcpc-vm2:~$ ls -l /usr/hdp/current/oozie-server
lrwxrwxrwx 1 root root 27 Nov 24 15:14 /usr/hdp/current/oozie-server -> /usr/hdp/2.2.0.0-2041/oozie
```

That is why the command with `#{OOZIE_CLIENT_PATH}` was commented out. I merely deleted the commented out code in this commit.

cbaenziger commented on cookbooks/bcpc-hadoop/recipes/oozie.rb:
```diff
@@ -181,9 +183,9 @@
+    status=`sudo -u oozie oozie admin -oozie #{OOZIE_URL} -status 2>&1` 
     while not /NORMAL/ =~ status and $?.to_i
-      status=`sudo -u oozie oozie admin -oozie http://"#{float_host(node[:fqdn])}":11000/oozie -status 2>&1` 
+      status=`sudo -u oozie oozie admin -oozie #{OOZIE_URL} -status 2>&1` 
```
Maybe create a function or some sort of definition for testing that Oozie's alive? This is the second instance of this status check code in the recipe.

cbaenziger commented on roles/BCPC-Hadoop-Head-MapReduce.json:
```diff
@@ -1,6 +1,13 @@
+    "bcpc": { 
+      "hadoop": {
+        "oozie": {
+          "sharelib_checksum": "updated_automatically"
```
I would prefer for `oozie.rb` to handle this attribute being unset (or having a cookbook default attribute) rather than putting a required piece of info in a role. If someone included the Oozie recipe they would also need to set this in some one-off way.

kiiranh commented:
Code changes have been tested with kerberos enabled as well. 

**Note**:
Current code only [sets](https://github.com/kiiranh/chef-bach/blob/oozie_sharelib_fix/cookbooks/bcpc-hadoop/recipes/oozie.rb#L115-L116) the sharelib file checksum in the node object but does not perform a `node.save`. It expects that the attribute will be persisted when the node object is saved at the end of the chef-client run. Caveat here is: it could take extra chef-client runs for this code to converge if the initial chef-client runs fail to save the node object (e.g.: due to some exception later on in the run_list). You might end up with a couple extra sharelib directories in hdfs because of this. Those will be cleaned up per the sharelib cleanup schedule mentioned [here](http://blog.cloudera.com/blog/2014/05/how-to-use-the-sharelib-in-apache-oozie-cdh-5/)

cbaenziger commented:
Please provide function documentation for the `utils.rb` code. Per the `README.md` [Tomdoc](http://tomdoc.org/) style would be good. (See "Method Documentation" therein.)

cbaenziger commented on cookbooks/bcpc-hadoop/libraries/utils.rb:
```diff
@@ -395,3 +395,33 @@ def has_vip?
+      Chef::Log.info("Sharelibupdate: Failed to update sharelib on #{host}")
+    end
+  rescue Exception => e
+    Chef::Log.error(e.message)
```
I would like us not to swallow exceptions unless we have a good reason for handling them elsewhere.

kiiranh commented:
Incorporated review comments:
- Added documentation for the proposed library functions
- Increased timeout for shellout commands from 10 to 20 to factor in the retries (default 4 times, after intervals 1, 2, 4, 8 seconds) that the Oozie client makes to try to connect to Oozie server.
- Removed `rescue` blocks that swallowed exceptions because the failures we expect (oozie server not reachable/running, user oozie doesn't exist, user oozie doesn't have permission to execute the commands) will cause a non-zero exit status for shellout and will not throw any exceptions. So if and when an exception does occur we want it to be noticed.

pu239ppy commented on cookbooks/bcpc-hadoop/libraries/utils.rb:
```diff
@@ -395,3 +395,45 @@ def has_vip?
+    if cmd.exitstatus == 0
+      Chef::Log.info("Sharelibupdate: Updated sharelib on #{host}")
+    else
+      Chef::Log.info("Sharelibupdate: sharelibupdate command failed on #{host}")
```
Consider returning STDERR and STDOUT as part of your error message

kiiranh replied:
@pu239ppy Done

pu239ppy commented:
Looks good
